packages/flutter_tools/bin/xcode_backend.dart

```diff
@@ -0,0 +1,466 @@
+        embedFlutterFrameworks();
+        break;
+      case 'test_observatory_bonjour_service':
+        // Exposed for integration testing only.
+        addObservatoryBonjourService();
+    }
+  }
+
+  bool existsDir(String path) {
+    final Directory dir = Directory(path);
+    return dir.existsSync();
+  }
+
+  bool existsFile(String path) {
+    final File file = File(path);
+    return file.existsSync();
+  }
+
+  /// Run given command in a synchronous subprocess.
+  ///
+  /// Will throw [Exception] if the exit code is not 0.
+  ProcessResult runSync(
+    String bin,
+    List<String> args, {
+    bool verbose = false,
+    bool allowFail = false,
+    String? workingDirectory,
+  }) {
+    if (verbose) {
+      print('♦ $bin ${args.join(' ')}');
+    }
+    final ProcessResult result = Process.runSync(
+      bin,
+      args,
+      workingDirectory: workingDirectory,
+    );
+    echo(result.stdout as String);
+    if ((result.stderr as String).isNotEmpty) {
+      echoError(result.stderr as String);
+    }
+    if (!allowFail && result.exitCode != 0) {
+      stderr.write('${result.stderr}\n');
+      throw Exception(
+        'Command "$bin ${args.join(' ')}" exited with code ${result.exitCode}',
+      );
+    }
+    return result;
+  }
+
+  /// Log message to stderr.
+  void echoError(String message) {
+    stderr.write('$message\n');
+  }
+
+  /// Log message to stdout.
+  void echo(String message) {
+    print(message);
+  }
+
+  /// Exit the application with the given exit code.
+  ///
+  /// Exists to allow overriding in tests.
+  Never exitApp(int code) {
+    exit(code);
+  }
+
+  /// Return value from environment if it exists, else throw [Exception].
+  String environmentEnsure(String key) {
+    final String? value = environment[key];
+    if (value == null) {
+      throw Exception(
+        'Expected the environment variable "$key" to exist, but it did not!',
+      );
+    }
+    return value;
+  }
+
+  RandomAccessFile? scriptOutputStreamFile;
+
+  // When provided with a pipe by the host Flutter build process, output to the
+  // pipe goes to stdout of the Flutter build process directly.
+  void streamOutput(String output) {
+    scriptOutputStreamFile?.writeStringSync(output);
+  }
+
+
+  String parseFlutterBuildMode() {
+    // Use FLUTTER_BUILD_MODE if it's set, otherwise use the Xcode build configuration name
+    // This means that if someone wants to use an Xcode build config other than Debug/Profile/Release,
+    // they _must_ set FLUTTER_BUILD_MODE so we know what type of artifact to build.
+    final String? buildMode = (environment['FLUTTER_BUILD_MODE'] ?? environment['CONFIGURATION'])?.toLowerCase();
+    if (buildMode == null) {
+      throw Exception(
+        'could not determine buildMode from either FLUTTER_BUILD_MODE or '
+        'CONFIGURATION environment variables');
+    }
+
+    if (buildMode.contains('release')) {
+      return 'release';
+    }
```

Comment:
I think we should keep the scripts 1:1 for the first commit, then update behavior in subsequent commits.  This was a strict equality check in the old script (unless I'm missing some bash-ism).  Otherwise I think this needs a flavor integration test?

Reply:
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/bin/xcode_backend.sh#L48

```
  case "$build_mode" in
    *release*) build_mode="release";;
    *profile*) build_mode="profile";;
    *debug*) build_mode="debug";;
```

The asterisks are globs meaning anything

Reply:
Unless i'm missing an even deeper bash-ism

Reply:
> (unless I'm missing some bash-ism)